### PR TITLE
Margin template

### PR DESCRIPTION
**Pull request by Stefano80**
This PR is to initiate a discussion and get feedback. It is not intended to be merged. I ask Marco if it is ok to keep it open for some time in order for the community to contribute.

I noticed that all our margins (apart of razor) can be written in the form margin = a + b * d + c * d^2, where d is the depth.

I reformulated the code making margin a template using an external table of the coefficients. This could not be done for razoring margins, which constitute an exception. I left the razor margins as they are now for inspection, but added a commented line showing how this would look like. If there is interest in the community, I will try to tune all values at once and see whether it is possible to make this patch pass.

Fot this, https://github.com/official-stockfish/Stockfish/pull/933 is needed.

As it is now, this is a nonfunctional change.

I see some advantages in this format:
+ all pruning mechanisms based on margins are documented in the enum which is used to address the table.
+ all values are kept in a clean table, which makes maybe easier to recognize patterns
+ the code in which margins are used becomes more uniform over the search function and (in my opinion) more readable
+ the division by ONE_PLY is moved in a single place instead of being kept in dozen of different places

I also see some disadvantages:
- we add approx 10 LOCs
- it is unclear to me whether this way of using templates is sensible or not

### CODE REVIEW

File: src/search.cpp [hunk 1]
```diff
@@ -62,10 +62,22 @@ namespace {
 
   // Different node types, used as a template parameter
   enum NodeType { NonPV, PV };
+  enum MarginType {Razor, FutilityChild, FutilityParent, Null, SeeRoot, SeeNonRoot, MarginsNB = 6};
```
Comment (vondele) [inline src/search.cpp]: what about using NullMove instead of Null in this enum. Null is a loaded term.

File: src/search.cpp [hunk 2]
```diff
@@ -62,10 +62,22 @@ namespace {
...
+      { 500,   30,    0},  // Razor
+      { 0,    150,    0},  // Futility child
+      { 256,  200,    0},  // Futility parent
+      {-210,   35,    0},  // Static null
+      {   0, -248,    0},  // See at root
+      {   0,    0,  -35}   // See at non root
+  };
+
+  template <MarginType mt> Value margin(Depth d){
+    return Value((PruningMargins[mt][0] + d * PruningMargins[mt][1] + int(d) * d * PruningMargins[mt][2])/ONE_PLY);
```
Comment (Stefano80) [inline src/search.cpp]: Right, I fixed it in the version I am going to tune on the framework.

File: src/search.cpp (hunk 1 above)
Comment (Stefano80) [inline src/search.cpp]: Right, if eventually it is taken into consideration I will follow your suggestion.

**Comment by Stefano80** [comment]
To see what this patch could achieve, I incorporated the razor margins as proposed by @vondele, made it ONE_PLY invariant and submitted an SPSA test on the framework:

http://tests.stockfishchess.org/tests/view/585a5bc90ebc5903140c5fba

**Comment by lucasart** [comment]
I don't see the point of this patch. It makes the code more complex and less readable for no good reason. If you need to change the formula of razor margin, there's no need for this patch.

**Comment by Stefano80** [comment]
Hi @lucasart,

the point is to try to unify logic. At the moment, we have in search.cpp several places in which we compare an approximate position evaluation or gain with alpha or beta or 0 and we add some margin with a polynomial dependency on depth.

We do this essentially identical operation in several flavours:
- we address an int array in razoring
` eval + razor_margin[depth / ONE_PLY] <= alpha`
- we call a Value-valued function in futility pruning (child)
`eval - futility_margin(depth) >= beta`
- we just drop the value in place in null move search, futility pruning (parent) and SEE pruning at non root
`ss->staticEval >= beta - 35 * (depth / ONE_PLY - 6)`
`ss->staticEval + 256 + 200 * lmrDepth <= alpha`
`!pos.see_ge(move, Value(-35 * lmrDepth * lmrDepth))`
- we specify the margin in multiples of PawnValueEg in SEE pruning at root
`!pos.see_ge(move, -PawnValueEg * (depth / ONE_PLY))`

Btw, we use a completely different but equivalent formulation in ProbCut (not addresses by this patch, I just noticed it now).

At a superficial analysis, it could look like the code is more complex, since it uses templates. But at the moment we just define 6 different logical operations, and we do not even take advantage of templates. Also we have the problem, that we always have to divide by ONE_PLY to keep ONE_PLY in dependency (it could be a solution to just remove that, but @mcostalba wanted to keep it, so let us not discuss it here).

Ah, and by the way, you have the byproduct that you can tune everything at once.

Summarizing: no the code is not more complex in this patch. Readability is obviously better in the patch since we use the same semantic for all margins, and we do not do calculations in place.

**Comment by Stefano80** [comment]
So, my attempts did not go well. I also noticed that the template version is slightly slower than its plain counterpart. I close this.

**Pull request closed without merging**